<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing differential co-expression with csdR • csdR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing differential co-expression with csdR">
<meta property="og:description" content="csdR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">csdR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/csdR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AlmaasLab/csdR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="csdR_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing differential co-expression with csdR</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/AlmaasLab/csdR/blob/master/vignettes/csdR.Rmd"><code>vignettes/csdR.Rmd</code></a></small>
      <div class="hidden name"><code>csdR.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="when-and-why-to-use-this-package" class="section level1">
<h1 class="hasAnchor">
<a href="#when-and-why-to-use-this-package" class="anchor"></a>When and why to use this package</h1>
<p>The purpose of this package is to compare the gene expression of the genes in <em>two</em> different conditions. The most typical case is when comparing gene expression in patients with a disease with gene expression in study participants without the disease. Hence, we may construct a network containing genes which are relevant for the development of the disease. The input data may come from different measurements of expression such as microarray, proteomics or RNA-seq as long as:</p>
<ul>
<li>There are no missing values in the data. Consider filling in with average values or pseudo-values if this is not the case.</li>
<li>The expression values are coded as continuous numerical values which are comparable between samples. Note that only the ranks of each gene across the samples does matter as CSD uses the rank-based Spearman correlation.</li>
<li>The gene labels for the two conditions must match.</li>
</ul>
<p>For differential gene-expression involving more than two separate conditions, consider <code>CoDiNA</code> <span class="citation">(Morselli Gysi 2020)</span> instead.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>This package is hosted on Bioconductor. To install it, type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"csdR"</span><span class="op">)</span></code></pre></div>
<p>Then,</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://almaaslab.github.io/csdR">csdR</a></span><span class="op">)</span>
<span class="co">#&gt; </span></code></pre></div>
<p>should load the package into the current <strong>R</strong> session. For this vignette, we further load some auxiliary packages and set the random seed</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidyverse/glue">glue</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">45394534</span><span class="op">)</span></code></pre></div>
</div>
<div id="some-theoretical-considerations" class="section level1">
<h1 class="hasAnchor">
<a href="#some-theoretical-considerations" class="anchor"></a>Some theoretical considerations</h1>
<p>This is a re-implementation and slight modification of the CSD algorithm presented by Voigt <em>et al.</em> 2017<span class="citation">(Voigt 2017)</span>. In the first phase, the algorithm finds the co-expression between all genes within each condition using the Spearman correlation. For each pair of genes, we apply bootstrapping across the samples and compute the mean Spearman correlation <span class="math inline">\(\rho_1\)</span> and <span class="math inline">\(\rho_2\)</span> for the two conditions and the associated standard errors <span class="math inline">\(\sigma_1\)</span> and <span class="math inline">\(\sigma_2\)</span>. In the second stage, the values for the two conditions are compared and gives us the following <em>differential</em> co-expression scores:</p>
<ul>
<li>Conserved score, <span class="math inline">\(C=\frac{\left|\rho_1+\rho_2\right|}{\sqrt{\sigma_1^2+\sigma_2^2}}\)</span>, a high value indicating the same strong co-expression in both conditions.</li>
<li>Specific score, <span class="math inline">\(S=\frac{\left|\left|\rho_1\right|-\left|\rho_2\right|\right|}{\sqrt{\sigma_1^2+\sigma_2^2}}\)</span>, a high value indicating a strong co-expression in one condition, but not in the other.</li>
<li>Differentiated score, <span class="math inline">\(D=\frac{\left|\left|\rho_1\right|+\left|\rho_2\right|-\left|\rho_1+\rho_2\right|\right|}{\sqrt{\sigma_1^2+\sigma_2^2}}\)</span>, a high value indicating a strong co-expression in both conditions, but with the opposite sign.</li>
</ul>
</div>
<div id="workflow-outline" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-outline" class="anchor"></a>Workflow outline</h1>
<p>In this example, we are provided by two expression expression matrices from thyroid glands, <code>sick_expression</code> for patients with thyroid cancer and <code>normal_expression</code> for healthy controls. To run the CSD analysis for these two conditions, we simply do the following:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"sick_expression"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"normal_expression"</span><span class="op">)</span>
<span class="va">csd_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_csd.html">run_csd</a></span><span class="op">(</span>
    x_1 <span class="op">=</span> <span class="va">sick_expression</span>, x_2 <span class="op">=</span> <span class="va">normal_expression</span>,
    n_it <span class="op">=</span> <span class="fl">10</span>, nThreads <span class="op">=</span> <span class="fl">2L</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<p>After obtaining these results, we may write them to disk. However, for datasets with thousands of genes, we will get millions upon millions of gene pairs. Writing the results to disk is likely to fill up gigabytes of valuable storage space while the disk IO itself might take a considerable amount of time. Furthermore, we must reduce the information load to create meaningful results, so we better to that while the data is still in memory. We decide to select the 100 edges with highest C, S, and D-score.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pairs_to_pick</span> <span class="op">&lt;-</span> <span class="fl">100L</span>
<span class="va">c_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partial_argsort.html">partial_argsort</a></span><span class="op">(</span><span class="va">csd_results</span><span class="op">$</span><span class="va">cVal</span>, <span class="va">pairs_to_pick</span><span class="op">)</span>
<span class="va">c_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">c_filter</span>, <span class="op">]</span>
<span class="va">s_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partial_argsort.html">partial_argsort</a></span><span class="op">(</span><span class="va">csd_results</span><span class="op">$</span><span class="va">sVal</span>, <span class="va">pairs_to_pick</span><span class="op">)</span>
<span class="va">s_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">s_filter</span>, <span class="op">]</span>
<span class="va">d_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partial_argsort.html">partial_argsort</a></span><span class="op">(</span><span class="va">csd_results</span><span class="op">$</span><span class="va">dVal</span>, <span class="va">pairs_to_pick</span><span class="op">)</span>
<span class="va">d_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">d_filter</span>, <span class="op">]</span></code></pre></div>
<p>Why does the <code>csdR</code> package provide a general <code>partial_argsort</code> function which takes in a numeric vector and spits out the indecies of the largest elements instead of a more specialized function directly extracting the top results from the dataframe? The answer is flexibility. Writing an additional line of code and a dollar sign is not that much work after all and we may want more flexible approaches such as displaying the union of the C, S- and D-edges:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">csd_filter</span> <span class="op">&lt;-</span> <span class="va">c_filter</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/union.html">union</a></span><span class="op">(</span><span class="va">s_filter</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/union.html">union</a></span><span class="op">(</span><span class="va">d_filter</span><span class="op">)</span>
<span class="va">csd_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">csd_filter</span>, <span class="op">]</span></code></pre></div>
<div id="how-to-we-approach-from-here" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-we-approach-from-here" class="anchor"></a>How to we approach from here?</h2>
<p>The next logical step is to construct a network and do some analysis. This is outside the scope of this package, but we will provide some pointers for completeness. One viable approach is to use the ordinary <code>write.table</code> function to write the results of a file and then use an external tools such as Cytoscape to further make conclusions. Often, you may want to make an ontology enrichment of the genes.</p>
<p>The other option is of course to continue using R. Here, we provide an example of combining the C-, S- and D-networks and coloring the edges blue, green and red, respectively depending of where they come from.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">c_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">c_frame</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">s_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">s_frame</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">d_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">d_frame</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html">E</a></span><span class="op">(</span><span class="va">c_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="st">"C"</span>
<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html">E</a></span><span class="op">(</span><span class="va">s_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="st">"S"</span>
<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html">E</a></span><span class="op">(</span><span class="va">d_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="st">"D"</span>
<span class="va">combined_network</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/union.html">union</a></span><span class="op">(</span><span class="va">c_network</span>, <span class="va">s_network</span>, <span class="va">d_network</span><span class="op">)</span>
<span class="co"># Auxillary function for combining</span>
<span class="co"># the attributes of the three networks in a proper way</span>
<span class="va">join_attributes</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">graph</span>, <span class="va">attribute</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span>
        test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"{attribute}_1"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
        yes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span>
            test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"{attribute}_2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
            yes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"{attribute}_3"</span><span class="op">)</span><span class="op">)</span>,
            no <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"{attribute}_2"</span><span class="op">)</span><span class="op">)</span>
        <span class="op">)</span>,
        no <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"{attribute}_1"</span><span class="op">)</span><span class="op">)</span>
    <span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html">E</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="fu">join_attributes</span><span class="op">(</span><span class="va">combined_network</span>, <span class="st">"edge_type"</span><span class="op">)</span>
<span class="va">layout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/layout_nicely.html">layout_nicely</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html">E</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html">E</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span>,
    C <span class="op">=</span> <span class="st">"darkblue"</span>, S <span class="op">=</span> <span class="st">"green"</span>, D <span class="op">=</span> <span class="st">"darkred"</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">combined_network</span>, layout <span class="op">=</span> <span class="va">layout</span>,
    vertex.size <span class="op">=</span> <span class="fl">3</span>, edge.width <span class="op">=</span> <span class="fl">2</span>, vertex.label.cex <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></code></pre></div>
<p><img src="csdR_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
</div>
<div id="considerations-to-note" class="section level1">
<h1 class="hasAnchor">
<a href="#considerations-to-note" class="anchor"></a>Considerations to note</h1>
<div id="number-of-bootstrap-iterations" class="section level2">
<h2 class="hasAnchor">
<a href="#number-of-bootstrap-iterations" class="anchor"></a>Number of bootstrap iterations</h2>
<p>As with any bootstrap procedure the number of iterations represented by the argument <code>n_it</code> needs to be <em>sufficiently large</em> in order to get reproducible results. What this means is a matter of trial and error. In general this means that you should re-run the computations with a different random seed to see whether the number of bootstrap iterations are sufficient. Experience has shown that ~ 100 iterations might be sufficient to reproduce almost the same results in some cases, whereas in other cases, especially when the values are close, you may choose to run several thousand iterations.</p>
</div>
<div id="memory-consumption" class="section level2">
<h2 class="hasAnchor">
<a href="#memory-consumption" class="anchor"></a>Memory consumption</h2>
<p>For datasets with 20 000 to 30 000 genes, a considerable amount of memory is consumed during the computations. It it therefore not recommended in such cases to run CSD on your laptop or even a workstation, but rather on a compute server with several hundreds GB of RAM.</p>
</div>
<div id="number-of-top-gene-pairs-to-pick" class="section level2">
<h2 class="hasAnchor">
<a href="#number-of-top-gene-pairs-to-pick" class="anchor"></a>Number of top gene pairs to pick</h2>
<p>How many gene pairs to select depends on the specific needs and how big a network you want to handle. A 10 000 edge network may not be easy to visualize, but quantitative network metrics can still be extracted. Also, generating more edges than necessary usually does not make any major harm as superfluous edges can quickly be filter out afterwards. However, if you select fewer edges than you actually need, you have to re-do all calculations to increase the number.</p>
</div>
</div>
<div id="session-info-for-this-vignette" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info-for-this-vignette" class="anchor"></a>Session info for this vignette</h1>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=nb_NO.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=nb_NO.UTF-8        LC_COLLATE=nb_NO.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=nb_NO.UTF-8    LC_MESSAGES=nb_NO.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=nb_NO.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=nb_NO.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] dplyr_1.0.7      glue_1.5.0       igraph_1.2.8     magrittr_2.0.1  </span>
<span class="co">#&gt; [5] csdR_1.1.0       BiocStyle_2.20.2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] bitops_1.0-7            matrixStats_0.61.0-9000 fs_1.5.0               </span>
<span class="co">#&gt;   [4] bit64_4.0.5             RColorBrewer_1.1-2      doParallel_1.0.16      </span>
<span class="co">#&gt;   [7] httr_1.4.2              rprojroot_2.0.2         GenomeInfoDb_1.28.1    </span>
<span class="co">#&gt;  [10] dynamicTreeCut_1.63-1   backports_1.3.0         tools_4.1.2            </span>
<span class="co">#&gt;  [13] bslib_0.3.1             utf8_1.2.2              R6_2.5.1               </span>
<span class="co">#&gt;  [16] rpart_4.1-15            Hmisc_4.5-0             DBI_1.1.1              </span>
<span class="co">#&gt;  [19] BiocGenerics_0.38.0     colorspace_2.0-2        nnet_7.3-16            </span>
<span class="co">#&gt;  [22] gridExtra_2.3           tidyselect_1.1.1        preprocessCore_1.54.0  </span>
<span class="co">#&gt;  [25] bit_4.0.4               compiler_4.1.2          WGCNA_1.70-3           </span>
<span class="co">#&gt;  [28] textshaping_0.3.5       Biobase_2.52.0          htmlTable_2.2.1        </span>
<span class="co">#&gt;  [31] desc_1.4.0              bookdown_0.24           sass_0.4.0             </span>
<span class="co">#&gt;  [34] checkmate_2.0.0         scales_1.1.1            pkgdown_1.6.1          </span>
<span class="co">#&gt;  [37] systemfonts_1.0.3       stringr_1.4.0           digest_0.6.28          </span>
<span class="co">#&gt;  [40] foreign_0.8-81          rmarkdown_2.9           XVector_0.32.0         </span>
<span class="co">#&gt;  [43] RhpcBLASctl_0.20-137    jpeg_0.1-8.1            base64enc_0.1-3        </span>
<span class="co">#&gt;  [46] pkgconfig_2.0.3         htmltools_0.5.2         highr_0.9              </span>
<span class="co">#&gt;  [49] fastmap_1.1.0           htmlwidgets_1.5.4       rlang_0.4.12           </span>
<span class="co">#&gt;  [52] impute_1.66.0           rstudioapi_0.13         RSQLite_2.2.7          </span>
<span class="co">#&gt;  [55] jquerylib_0.1.4         generics_0.1.1          jsonlite_1.7.2         </span>
<span class="co">#&gt;  [58] RCurl_1.98-1.3          GO.db_3.10.0            GenomeInfoDbData_1.2.6 </span>
<span class="co">#&gt;  [61] Formula_1.2-4           Matrix_1.3-4            Rcpp_1.0.7             </span>
<span class="co">#&gt;  [64] munsell_0.5.0           S4Vectors_0.30.0        fansi_0.5.0            </span>
<span class="co">#&gt;  [67] lifecycle_1.0.1         stringi_1.7.5           yaml_2.2.1             </span>
<span class="co">#&gt;  [70] zlibbioc_1.38.0         grid_4.1.2              blob_1.2.2             </span>
<span class="co">#&gt;  [73] parallel_4.1.2          crayon_1.4.2            lattice_0.20-45        </span>
<span class="co">#&gt;  [76] Biostrings_2.60.1       splines_4.1.2           KEGGREST_1.32.0        </span>
<span class="co">#&gt;  [79] knitr_1.36              pillar_1.6.4            fastcluster_1.2.3      </span>
<span class="co">#&gt;  [82] codetools_0.2-18        stats4_4.1.2            evaluate_0.14          </span>
<span class="co">#&gt;  [85] latticeExtra_0.6-29     data.table_1.14.2       BiocManager_1.30.16    </span>
<span class="co">#&gt;  [88] png_0.1-7               vctrs_0.3.8             foreach_1.5.1          </span>
<span class="co">#&gt;  [91] gtable_0.3.0            purrr_0.3.4             assertthat_0.2.1       </span>
<span class="co">#&gt;  [94] cachem_1.0.6            ggplot2_3.3.5           xfun_0.24              </span>
<span class="co">#&gt;  [97] ragg_1.1.3              survival_3.2-13         tibble_3.1.6           </span>
<span class="co">#&gt; [100] iterators_1.0.13        AnnotationDbi_1.54.1    memoise_2.0.0          </span>
<span class="co">#&gt; [103] IRanges_2.26.0          cluster_2.1.2           ellipsis_0.3.2</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-MorselliGysi2020" class="csl-entry">
Morselli Gysi, Tiago AND Zebardast, Deisy AND de Miranda Fragoso. 2020. <span>“Whole Transcriptomic Network Analysis Using Co-Expression Differential Network Analysis (CoDiNA).”</span> <em>PLOS ONE</em> 15 (10): 1–28. <a href="https://doi.org/10.1371/journal.pone.0240523">https://doi.org/10.1371/journal.pone.0240523</a>.
</div>
<div id="ref-Voigt2017" class="csl-entry">
Voigt, Katja AND Almaas, André AND Nowick. 2017. <span>“A Composite Network of Conserved and Tissue Specific Gene Interactions Reveals Possible Genetic Interactions in Glioma.”</span> <em>PLOS Computational Biology</em> 13 (9): 1–34. <a href="https://doi.org/10.1371/journal.pcbi.1005739">https://doi.org/10.1371/journal.pcbi.1005739</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jakob Peder Pettersen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
