<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing differential co-expression with csdR • csdR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing differential co-expression with csdR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">csdR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.13.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/csdR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AlmaasLab/csdR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing differential co-expression with
csdR</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/AlmaasLab/csdR/blob/main/vignettes/csdR.Rmd" class="external-link"><code>vignettes/csdR.Rmd</code></a></small>
      <div class="hidden name"><code>csdR.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="when-and-why-to-use-this-package">When and why to use this package<a class="anchor" aria-label="anchor" href="#when-and-why-to-use-this-package"></a>
</h2>
<p>The purpose of this package is to compare the gene expression of the
genes in <em>two</em> different conditions. The most typical case is
when comparing gene expression in patients with a disease with gene
expression in study participants without the disease. Hence, we may
construct a network containing genes which are relevant for the
development of the disease. The input data may come from different
measurements of expression such as microarray, proteomics or RNA-seq as
long as:</p>
<ul>
<li>There are no missing values in the data. Consider filling in with
average values or pseudo-values if this is not the case.</li>
<li>The expression values are coded as continuous numerical values which
are comparable between samples. Note that only the ranks of each gene
across the samples does matter as CSD uses the rank-based Spearman
correlation.</li>
<li>The gene labels for the two conditions must match.</li>
</ul>
<p>For differential gene-expression involving more than two separate
conditions, consider <code>CoDiNA</code> <span class="citation">(Morselli Gysi 2020)</span> instead.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>This package is hosted on Bioconductor. To install it, type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"csdR"</span><span class="op">)</span></span></code></pre></div>
<p>Then,</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://almaaslab.github.io/csdR">csdR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
<p>should load the package into the current <strong>R</strong> session.
For this vignette, we further load some auxiliary packages and set the
random seed</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">45394534</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="some-theoretical-considerations">Some theoretical considerations<a class="anchor" aria-label="anchor" href="#some-theoretical-considerations"></a>
</h2>
<p>This is a re-implementation and slight modification of the CSD
algorithm presented by Voigt <em>et al.</em> 2017<span class="citation">(Voigt 2017)</span>. In the first phase, the algorithm
finds the co-expression between all genes within each condition using
the Spearman correlation. For each pair of genes, we apply bootstrapping
across the samples and compute the mean Spearman correlation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ρ</mi><mn>1</mn></msub><annotation encoding="application/x-tex">\rho_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ρ</mi><mn>2</mn></msub><annotation encoding="application/x-tex">\rho_2</annotation></semantics></math>
for the two conditions and the associated standard errors
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>σ</mi><mn>1</mn></msub><annotation encoding="application/x-tex">\sigma_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>σ</mi><mn>2</mn></msub><annotation encoding="application/x-tex">\sigma_2</annotation></semantics></math>.
In the second stage, the values for the two conditions are compared and
gives us the following <em>differential</em> co-expression scores:</p>
<ul>
<li>Conserved score,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mo>=</mo><mfrac><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>ρ</mi><mn>1</mn></msub><mo>+</mo><msub><mi>ρ</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><msqrt><mrow><msubsup><mi>σ</mi><mn>1</mn><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mn>2</mn><mn>2</mn></msubsup></mrow></msqrt></mfrac></mrow><annotation encoding="application/x-tex">C=\frac{\left|\rho_1+\rho_2\right|}{\sqrt{\sigma_1^2+\sigma_2^2}}</annotation></semantics></math>,
a high value indicating the same strong co-expression in both
conditions.</li>
<li>Specific score,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mo>=</mo><mfrac><mrow><mo stretchy="true" form="prefix">|</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>ρ</mi><mn>1</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>−</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>ρ</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">|</mo></mrow><msqrt><mrow><msubsup><mi>σ</mi><mn>1</mn><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mn>2</mn><mn>2</mn></msubsup></mrow></msqrt></mfrac></mrow><annotation encoding="application/x-tex">S=\frac{\left|\left|\rho_1\right|-\left|\rho_2\right|\right|}{\sqrt{\sigma_1^2+\sigma_2^2}}</annotation></semantics></math>,
a high value indicating a strong co-expression in one condition, but not
in the other.</li>
<li>Differentiated score,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mo>=</mo><mfrac><mrow><mo stretchy="true" form="prefix">|</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>ρ</mi><mn>1</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>+</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>ρ</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>−</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>ρ</mi><mn>1</mn></msub><mo>+</mo><msub><mi>ρ</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">|</mo></mrow><msqrt><mrow><msubsup><mi>σ</mi><mn>1</mn><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mn>2</mn><mn>2</mn></msubsup></mrow></msqrt></mfrac></mrow><annotation encoding="application/x-tex">D=\frac{\left|\left|\rho_1\right|+\left|\rho_2\right|-\left|\rho_1+\rho_2\right|\right|}{\sqrt{\sigma_1^2+\sigma_2^2}}</annotation></semantics></math>,
a high value indicating a strong co-expression in both conditions, but
with the opposite sign.</li>
</ul>
</div>
<div class="section level2">
<h2 id="workflow-outline">Workflow outline<a class="anchor" aria-label="anchor" href="#workflow-outline"></a>
</h2>
<p>In this example, we are provided by two expression expression
matrices from thyroid glands, <code>sick_expression</code> for patients
with thyroid cancer and <code>normal_expression</code> for healthy
controls. To run the CSD analysis for these two conditions, we simply do
the following:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sick_expression"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"normal_expression"</span><span class="op">)</span></span>
<span><span class="va">csd_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_csd.html">run_csd</a></span><span class="op">(</span></span>
<span>    x_1 <span class="op">=</span> <span class="va">sick_expression</span>, x_2 <span class="op">=</span> <span class="va">normal_expression</span>,</span>
<span>    n_it <span class="op">=</span> <span class="fl">10</span>, nThreads <span class="op">=</span> <span class="fl">2L</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>After obtaining these results, we may write them to disk. However,
for datasets with thousands of genes, we will get millions upon millions
of gene pairs. Writing the results to disk is likely to fill up
gigabytes of valuable storage space while the disk IO itself might take
a considerable amount of time. Furthermore, we must reduce the
information load to create meaningful results, so we better to that
while the data is still in memory. We decide to select the 100 edges
with highest C, S, and D-score.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairs_to_pick</span> <span class="op">&lt;-</span> <span class="fl">100L</span></span>
<span><span class="va">c_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partial_argsort.html">partial_argsort</a></span><span class="op">(</span><span class="va">csd_results</span><span class="op">$</span><span class="va">cVal</span>, <span class="va">pairs_to_pick</span><span class="op">)</span></span>
<span><span class="va">c_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">c_filter</span>, <span class="op">]</span></span>
<span><span class="va">s_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partial_argsort.html">partial_argsort</a></span><span class="op">(</span><span class="va">csd_results</span><span class="op">$</span><span class="va">sVal</span>, <span class="va">pairs_to_pick</span><span class="op">)</span></span>
<span><span class="va">s_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">s_filter</span>, <span class="op">]</span></span>
<span><span class="va">d_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partial_argsort.html">partial_argsort</a></span><span class="op">(</span><span class="va">csd_results</span><span class="op">$</span><span class="va">dVal</span>, <span class="va">pairs_to_pick</span><span class="op">)</span></span>
<span><span class="va">d_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">d_filter</span>, <span class="op">]</span></span></code></pre></div>
<p>Why does the <code>csdR</code> package provide a general
<code>partial_argsort</code> function which takes in a numeric vector
and spits out the indecies of the largest elements instead of a more
specialized function directly extracting the top results from the
dataframe? The answer is flexibility. Writing an additional line of code
and a dollar sign is not that much work after all and we may want more
flexible approaches such as displaying the union of the C, S- and
D-edges:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">csd_filter</span> <span class="op">&lt;-</span> <span class="va">c_filter</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://r.igraph.org/reference/union.html" class="external-link">union</a></span><span class="op">(</span><span class="va">s_filter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://r.igraph.org/reference/union.html" class="external-link">union</a></span><span class="op">(</span><span class="va">d_filter</span><span class="op">)</span></span>
<span><span class="va">csd_frame</span> <span class="op">&lt;-</span> <span class="va">csd_results</span><span class="op">[</span><span class="va">csd_filter</span>, <span class="op">]</span></span></code></pre></div>
<div class="section level3">
<h3 id="how-to-we-approach-from-here">How to we approach from here?<a class="anchor" aria-label="anchor" href="#how-to-we-approach-from-here"></a>
</h3>
<p>The next logical step is to construct a network and do some analysis.
This is outside the scope of this package, but we will provide some
pointers for completeness. One viable approach is to use the ordinary
<code>write.table</code> function to write the results of a file and
then use an external tools such as Cytoscape to further make
conclusions. Often, you may want to make an ontology enrichment of the
genes.</p>
<p>The other option is of course to continue using R. Here, we provide
an example of combining the C-, S- and D-networks and coloring the edges
blue, green and red, respectively depending of where they come from.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span><span class="va">c_frame</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">s_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span><span class="va">s_frame</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">d_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span><span class="va">d_frame</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">c_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="st">"C"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">s_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="st">"S"</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">d_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="st">"D"</span></span>
<span><span class="va">combined_network</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/union.html" class="external-link">union</a></span><span class="op">(</span><span class="va">c_network</span>, <span class="va">s_network</span>, <span class="va">d_network</span><span class="op">)</span></span>
<span><span class="co"># Auxillary function for combining</span></span>
<span><span class="co"># the attributes of the three networks in a proper way</span></span>
<span><span class="va">join_attributes</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">graph</span>, <span class="va">attribute</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>        test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/edge_attr.html" class="external-link">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{attribute}_1"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        yes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>            test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/edge_attr.html" class="external-link">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{attribute}_2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            yes <span class="op">=</span> <span class="fu"><a href="https://r.igraph.org/reference/edge_attr.html" class="external-link">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{attribute}_3"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            no <span class="op">=</span> <span class="fu"><a href="https://r.igraph.org/reference/edge_attr.html" class="external-link">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{attribute}_2"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>        no <span class="op">=</span> <span class="fu"><a href="https://r.igraph.org/reference/edge_attr.html" class="external-link">edge_attr</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{attribute}_1"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span> <span class="op">&lt;-</span> <span class="fu">join_attributes</span><span class="op">(</span><span class="va">combined_network</span>, <span class="st">"edge_type"</span><span class="op">)</span></span>
<span><span class="va">layout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/layout_nicely.html" class="external-link">layout_nicely</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">combined_network</span><span class="op">)</span><span class="op">$</span><span class="va">edge_type</span>,</span>
<span>    C <span class="op">=</span> <span class="st">"darkblue"</span>, S <span class="op">=</span> <span class="st">"green"</span>, D <span class="op">=</span> <span class="st">"darkred"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">combined_network</span>, layout <span class="op">=</span> <span class="va">layout</span>,</span>
<span>    vertex.size <span class="op">=</span> <span class="fl">3</span>, edge.width <span class="op">=</span> <span class="fl">2</span>, vertex.label.cex <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span></code></pre></div>
<p><img src="csdR_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="considerations-to-note">Considerations to note<a class="anchor" aria-label="anchor" href="#considerations-to-note"></a>
</h2>
<div class="section level3">
<h3 id="number-of-bootstrap-iterations">Number of bootstrap iterations<a class="anchor" aria-label="anchor" href="#number-of-bootstrap-iterations"></a>
</h3>
<p>As with any bootstrap procedure the number of iterations represented
by the argument <code>n_it</code> needs to be <em>sufficiently
large</em> in order to get reproducible results. What this means is a
matter of trial and error. In general this means that you should re-run
the computations with a different random seed to see whether the number
of bootstrap iterations are sufficient. Experience has shown that ~ 100
iterations might be sufficient to reproduce almost the same results in
some cases, whereas in other cases, especially when the values are
close, you may choose to run several thousand iterations.</p>
</div>
<div class="section level3">
<h3 id="number-of-threads">Number of threads<a class="anchor" aria-label="anchor" href="#number-of-threads"></a>
</h3>
<p>The parallelization of <code>csdR</code> occurs within each
individual iteration. While parallelizing across the iterations would in
theory yield better CPU utilization, this approach is unfeasible due the
fact that memory consumption use would almost be proportional to the
number of threads. Instead, parallelization is used in computing the
co-expression matrix and updating the mean and variance for each
iteration. As shown in the <code>csdR</code> paper, parallelism can
provide 2x-3x speedup, but there are diminishing returns in the
performance gained, so running more than 10 threads is most likely a
waste a resources. Limited memory bandwidth is the most likely culprit
behind the failure to scale, so systems with higher memory bandwidth are
more likely to utilize multiple threads better.</p>
</div>
<div class="section level3">
<h3 id="memory-consumption">Memory consumption<a class="anchor" aria-label="anchor" href="#memory-consumption"></a>
</h3>
<p>For datasets with 20 000 to 30 000 genes, a considerable amount of
memory is consumed during the computations. It it therefore not
recommended in such cases to run CSD on your laptop or even a
workstation, but rather on a compute server with several hundreds GB of
RAM.</p>
</div>
<div class="section level3">
<h3 id="number-of-top-gene-pairs-to-pick">Number of top gene pairs to pick<a class="anchor" aria-label="anchor" href="#number-of-top-gene-pairs-to-pick"></a>
</h3>
<p>How many gene pairs to select depends on the specific needs and how
big a network you want to handle. A 10 000 edge network may not be easy
to visualize, but quantitative network metrics can still be extracted.
Also, generating more edges than necessary usually does not make any
major harm as superfluous edges can quickly be filter out afterwards.
However, if you select fewer edges than you actually need, you have to
re-do all calculations to increase the number.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info-for-this-vignette">Session info for this vignette<a class="anchor" aria-label="anchor" href="#session-info-for-this-vignette"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4      glue_1.8.0       igraph_2.1.4     magrittr_2.0.3  </span></span>
<span><span class="co">#&gt; [5] csdR_1.13.3      BiocStyle_2.34.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] DBI_1.2.3               gridExtra_2.3           rlang_1.1.5            </span></span>
<span><span class="co">#&gt;  [4] matrixStats_1.5.0       compiler_4.4.3          RSQLite_2.3.9          </span></span>
<span><span class="co">#&gt;  [7] png_0.1-8               systemfonts_1.2.1       vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [10] stringr_1.5.1           pkgconfig_2.0.3         crayon_1.5.3           </span></span>
<span><span class="co">#&gt; [13] fastmap_1.2.0           backports_1.5.0         XVector_0.46.0         </span></span>
<span><span class="co">#&gt; [16] rmarkdown_2.29          UCSC.utils_1.2.0        preprocessCore_1.68.0  </span></span>
<span><span class="co">#&gt; [19] ragg_1.3.3              bit_4.6.0               xfun_0.51              </span></span>
<span><span class="co">#&gt; [22] zlibbioc_1.52.0         cachem_1.1.0            GenomeInfoDb_1.42.3    </span></span>
<span><span class="co">#&gt; [25] jsonlite_1.9.1          blob_1.2.4              parallel_4.4.3         </span></span>
<span><span class="co">#&gt; [28] cluster_2.1.8           R6_2.6.1                bslib_0.9.0            </span></span>
<span><span class="co">#&gt; [31] stringi_1.8.4           rpart_4.1.24            jquerylib_0.1.4        </span></span>
<span><span class="co">#&gt; [34] Rcpp_1.0.14             bookdown_0.42           iterators_1.0.14       </span></span>
<span><span class="co">#&gt; [37] knitr_1.50              WGCNA_1.73              base64enc_0.1-3        </span></span>
<span><span class="co">#&gt; [40] IRanges_2.40.1          Matrix_1.7-2            splines_4.4.3          </span></span>
<span><span class="co">#&gt; [43] nnet_7.3-20             tidyselect_1.2.1        rstudioapi_0.17.1      </span></span>
<span><span class="co">#&gt; [46] yaml_2.3.10             doParallel_1.0.17       codetools_0.2-20       </span></span>
<span><span class="co">#&gt; [49] lattice_0.22-6          tibble_3.2.1            Biobase_2.66.0         </span></span>
<span><span class="co">#&gt; [52] KEGGREST_1.46.0         evaluate_1.0.3          foreign_0.8-88         </span></span>
<span><span class="co">#&gt; [55] desc_1.4.3              survival_3.8-3          Biostrings_2.74.1      </span></span>
<span><span class="co">#&gt; [58] pillar_1.10.1           BiocManager_1.30.25     checkmate_2.3.2        </span></span>
<span><span class="co">#&gt; [61] foreach_1.5.2           stats4_4.4.3            generics_0.1.3         </span></span>
<span><span class="co">#&gt; [64] S4Vectors_0.44.0        ggplot2_3.5.1           munsell_0.5.1          </span></span>
<span><span class="co">#&gt; [67] scales_1.3.0            RhpcBLASctl_0.23-42     Hmisc_5.2-3            </span></span>
<span><span class="co">#&gt; [70] tools_4.4.3             data.table_1.17.0       fs_1.6.5               </span></span>
<span><span class="co">#&gt; [73] fastcluster_1.2.6       grid_4.4.3              impute_1.80.0          </span></span>
<span><span class="co">#&gt; [76] AnnotationDbi_1.68.0    colorspace_2.1-1        GenomeInfoDbData_1.2.13</span></span>
<span><span class="co">#&gt; [79] htmlTable_2.4.3         Formula_1.2-5           cli_3.6.4              </span></span>
<span><span class="co">#&gt; [82] textshaping_1.0.0       gtable_0.3.6            dynamicTreeCut_1.63-1  </span></span>
<span><span class="co">#&gt; [85] sass_0.4.9              digest_0.6.37           BiocGenerics_0.52.0    </span></span>
<span><span class="co">#&gt; [88] htmlwidgets_1.6.4       memoise_2.0.1           htmltools_0.5.8.1      </span></span>
<span><span class="co">#&gt; [91] pkgdown_2.1.1           lifecycle_1.0.4         httr_1.4.7             </span></span>
<span><span class="co">#&gt; [94] GO.db_3.20.0            bit64_4.6.0-1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-MorselliGysi2020" class="csl-entry">
Morselli Gysi, Tiago AND Zebardast, Deisy AND de Miranda Fragoso. 2020.
<span>“Whole Transcriptomic Network Analysis Using Co-Expression
Differential Network Analysis (CoDiNA).”</span> <em>PLOS ONE</em> 15
(10): 1–28. <a href="https://doi.org/10.1371/journal.pone.0240523" class="external-link">https://doi.org/10.1371/journal.pone.0240523</a>.
</div>
<div id="ref-Voigt2017" class="csl-entry">
Voigt, Katja AND Almaas, André AND Nowick. 2017. <span>“A Composite
Network of Conserved and Tissue Specific Gene Interactions Reveals
Possible Genetic Interactions in Glioma.”</span> <em>PLOS Computational
Biology</em> 13 (9): 1–34. <a href="https://doi.org/10.1371/journal.pcbi.1005739" class="external-link">https://doi.org/10.1371/journal.pcbi.1005739</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jakob Peder Pettersen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
